Sprite Distort

Francessco121 opened the pull request:
This PR implements the sprite distortion effect as part of the sprite3 renderer.

I'm still fairly new to a lot of stuff in this project, so if there's anything I need to refactor or clean up please let me know!

The sprite distort effect needs to read the active framebuffer to do its thing. I implemented this by creating a separate framebuffer managed by the sprite renderer. When it comes time to render sprite distortion, the default framebuffer's color buffer is blitted (using `glBlitFramebuffer`) to this framebuffer. Then the color buffer of this framebuffer is bound to the shader as a texture. I'm not sure what the performance implications of this are. I have a feeling there's a faster way to implement this effect, but this was a super simple approach.

Since the sprite renderer managed framebuffer needs to have the same dimensions as the default framebuffer, I added the pixel width/height of the window to `SharedRenderState`. I'm not sure if this is the most appropriate way to pass this information to the renderer. Currently, the renderer will resize the framebuffer on demand and checks to see if the width/height changed each frame.

**Issues:**
~At the moment the effect is not perfectly accurate to the original game. Specifically, the texture coordinates used for sampling the framebuffer texture don't look right unless the Y-axis is flipped and then offset by \~0.06. I know this isn't right but I don't really understand why this is necessary.~

~Additionally, the current implementation issues a lot of draw calls (one for each sprite).~

water111 commented on docs/scratch/sprite_distort_vu1.txt:
```diff
@@ -0,0 +1,94 @@
+  ibne vi00, vi01, L1        |  nop                         ;; Loop for remaining sprites
+  isub vi02, vi03, vi02      |  nop                      
+  nop                        |  nop :e                   
+  nop                        |  nop                      
```
thanks for including this documentation!

water111 commented on game/graphics/opengl_renderer/OpenGLRenderer.cpp:
```diff
@@ -414,6 +414,9 @@ void OpenGLRenderer::setup_frame(int window_width_px,
   glDisable(GL_BLEND);
+
+  m_render_state.window_width_px = window_width_px;
+  m_render_state.window_height_px = window_height_px;
```
I think I'm running into issues related to this, when the game doesn't take up the full window:
![image](https://user-images.githubusercontent.com/48171810/178080243-79c78ed4-64dc-49b3-95c2-9d3c5bedcf61.png)

I'm not actually sure this is something that's supposed to be possible or not, but the resolution/aspect ratio settings don't really work for me in the first place.  This is probably a separate issue and shouldn't block this PR.

Francessco121 replied:
Turned out to be a simple issue with the framebuffer blit. The source dimensions needed to respect the XY offset of the renderer.

water111 commented on game/graphics/opengl_renderer/Sprite3.cpp:
```diff
@@ -116,15 +126,94 @@ Sprite3::Sprite3(const std::string& name, BucketId my_id)
+  );
+  glEnableVertexAttribArray(1);
+  glVertexAttribPointer(1,                                        // location 1 in the shader
+                        3,                                        // 2 floats per vert
```
This should be a 2 - it caused some interesting problems where the last vertex had messed up texture coordinates on my GPU:
![image](https://user-images.githubusercontent.com/48171810/178080418-0ea4b3eb-c1ab-424a-853b-a589d9c858f0.png)

Francessco121 replied:
Oops, good catch. I'm surprised that didn't cause issues on my machine...

water111 commented on game/graphics/opengl_renderer/Sprite3.cpp:
```diff
@@ -116,15 +126,94 @@ Sprite3::Sprite3(const std::string& name, BucketId my_id)
 void Sprite3::render_distorter(DmaFollower& dma,
                                SharedRenderState* render_state,
                                ScopedProfilerNode& prof) {
+  // First, make sure the distort framebuffer is the correct size
```
It might be nice to add a way to disable sprite distort, just in case it causes massive performance issues for some people.

You could follow the example of `m_3d_enable` to set up a checkbox in the debug GUI.

Francessco121 replied:
Done, also made it respect the global sprite `m_enabled` flag.

water111 commented on goal_src/jak1/engine/gfx/sprite/sprite-distort.gc:
```diff
@@ -111,92 +117,40 @@
-               )
+  
+  ;; set up GS registers
+  (dma-buffer-add-gs-set dma-buff
```
nice use of the macro

water111 reviewed (commented):
Looks great!  I had just a few comments, but other than those, I think this is good to go

water111 commented:
Also - to clarify - with the `st` fix and the "fit to screen" option, it worked perfectly for me, with very change in frame times

Francessco121 commented:
I addressed the comments and also did a bit of a refactor. Each 'stage' of the render (dma, setup, draw) are fully separate now and have their own profiling scopes.

Francessco121 commented:
The size of `sprite-aux-list` now scales up when `PC_BIG_MEMORY` is set just like the other sprite stuff. This fixes an issue mainly around the klaww fight and volcanic crater where most distort sprites wouldn't render since without particle sprite fade out the original 256 sprite cap is hit.

I think that's pretty much every issue solved! 😄

water111 commented on game/graphics/opengl_renderer/Sprite3.cpp:
```diff
@@ -116,43 +127,375 @@ Sprite3::Sprite3(const std::string& name, BucketId my_id)
+  }
+
+  if (m_distort_stats.total_tris == 0) {
+    // No distort sprites to draw, we can end early
```
good idea to add this, I imagine there's some places with no sprite distorts
